File: Algo/Leetcode/032LongestValidParentheses.cpp
// use stack
class Solution {
public:
    int longestValidParentheses(string s) {
        if(s.length() == 0) return 0;
        stack<int> st;
        st.push(-1);

        int max_len = 0;
        for(int i=0; i<s.length(); i++){
            if(s[i] == '(') st.push(i);
            else{
                st.pop();
                if(st.empty()) st.push(i);
                else max_len = max(max_len, i-st.top());
            }
        }
        return max_len;
    }
};


class Solution {
public:
    int longestValidParentheses(string s) {
        vector<int> f(s.length(),0);
        int res = 0;
        for(int i=s.length()-2;i>=0;i--){
            int match = i + f[i+1] + 1;
            if(s[i]=='(' && match < s.length() && s[match]==')'){
                f[i] = f[i+1] + 2;
                if(match+1<s.length()){
                    f[i]+=f[match+1];
                }  
            }
            res = max(res,f[i]);
        }
        return res;
    }
};

File: Algo/Leetcode/032LongestValidParentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        st = []
        n = len(s)
        dp = [0] * n
        for i, p in enumerate(s):
            if s[i] == '(':
                st.append((i, '('))
            else:
                if len(st) > 0 and st[-1][1] == '(':
                    m = st.pop()
                else:
                    st.append((i, ')'))

        while len(st) > 0:
            m = st.pop()
            dp[m[0]] = 1

        # find the longest 0 seq
        max_count = 0
        count = 0
        for m in dp:
            if m == 0:
                count += 1
                max_count = max(max_count, count)
            else:
                count = 0

        return max_count

